### PR TITLE
Use Char for single char strings

### PR DESCRIPTION
**Pull request by Sija**
Lil' change for noticeable gain:

```cr
require "benchmark"

Benchmark.ips do |x|
  io1 = IO::Memory.new
  io2 = IO::Memory.new

  x.report("<< Char") do
    io1 << 'a' << 'b' << 'c'
  end
  x.report("<< String") do
    io2 << "a" << "b" << "c"
  end
end
```

```
~ ❯ crystal run x.cr --release                                                                                                                                                                            
  << Char  47.06M ( 21.25ns) (±23.32%)       fastest
<< String  29.56M ( 33.83ns) (±17.99%)  1.59× slower

~ ❯ crystal run x.cr --release
  << Char  54.37M ( 18.39ns) (±31.44%)       fastest
<< String  28.09M ( 35.61ns) (±15.39%)  1.94× slower

~ ❯ crystal run x.cr --release
  << Char   48.6M ( 20.58ns) (±23.82%)       fastest
<< String  28.97M ( 34.51ns) (±17.46%)  1.68× slower
```

### CODE REVIEW

File: src/compiler/crystal/command/format.cr [hunk 1]
```diff
@@ -173,7 +173,7 @@ class Crystal::Command
       if check_files
         check_files << FormatResult.new(filename, FormatResult::Code::SYNTAX)
       else
-        STDERR << "Syntax Error:".colorize(:yellow).toggle(@color) << " " << ex.message << " at " << filename << ":" << ex.line_number << ":" << ex.column_number << "\n"
+        STDERR << "Syntax Error:".colorize(:yellow).toggle(@color) << " " << ex.message << " at " << filename << ":" << ex.line_number << ":" << ex.column_number << '\n'
```
Comment (bew) [inline src/compiler/crystal/command/format.cr]: You are missing some `" "` & `":"` here

Comment (Sija) [reply src/compiler/crystal/command/format.cr]: Yep, there are still some missing occurrences I intentionally left untouched due to compiler errors, which I didn't had time to debug, yet.

**Comment by Sija** [comment]
All tests have passed with an exception thrown?

```
Unhandled exception in spawn:

Error writing to socket: Broken pipe (Errno)
  from src/socket.cr:74:11 in 'unbuffered_write'
  from src/io/buffered.cr:112:14 in 'write'
  from src/io.cr:864:5 in 'write_byte'
  from src/io/buffered.cr:145:14 in 'write_byte'
  from src/char.cr:774:9 in 'to_s'
  from src/io.cr:197:5 in '<<'
  from src/http/client/response.cr:64:5 in 'to_io'
  from spec/std/http/client/client_spec.cr:14:9 in '->'
  from src/fiber.cr:255:3 in 'run'
  from src/fiber.cr:29:34 in '->'
  from ???
```

**Comment by oprypin** [comment]
I've seen that error in other PRs, probably unrelated

**Comment by Sija** [comment]
Ready for review 🚢

/cc @RX14 @asterite @ysbaddaden

**Comment by oprypin** [comment]
No :|
These could be doing different things, unfortunately

**Comment by RX14** [comment]
I actually find it interesting that `io << 'a'` is faster than `io << "a"`. `'a'` has to be converted to UTF-8 before it's memory copied, however `io << "a"` is already represented as UTF-8 by the compiler. I imagine it allows more compiler optimizations, being a scalar constant instead of a pointer constant.

I've just tested it with non-ASCII characters and strings win out :)

```
$ crystal run --release /tmp/bm.cr
<< String  45.19  ( 22.13ms) (± 6.91%)       fastest
  << Char  42.33  ( 23.63ms) (± 8.44%)  1.07× slower
```

```cr
require "benchmark"

Benchmark.ips do |x|
  io1 = IO::Memory.new
  io2 = IO::Memory.new

  x.report("<< String") do
    1_000_000.times do
      io2 << "҂" << "☢" << "☂"
    end
    io2 = IO::Memory.new
  end

  x.report("<< Char") do
    1_000_000.times do
      io1 << '҂' << '☢' << '☂'
    end
    io1 = IO::Memory.new
  end
end
```

**Comment by Sija** [comment]
@RX14 yes, but bear in mind that this PR changes **only** ASCII character strings, so benchmark posted in the description still stands.

**Comment by Sija** [comment]
> 'a' has to be converted to UTF-8 before it's memory copied

that's not the case, see `Char#to_s(io)`: https://github.com/crystal-lang/crystal/blob/5cd78fa51284d22dcf5b485b0939e790d2f54f54/src/char.cr#L770-L775

**Comment by RX14** [comment]
@Sija yes I know it's faster and why it's faster, but I thought it was interesting that Char was only faster than String when it could apply the single-bye ASCII optimization. I can see people taking "char is faster than string" further than makes sense...

**Comment by Sija** [comment]
@RX14 will it merge?

**Comment by r00ster91** [comment]
In [colorize.cr](https://github.com/Sija/crystal/blob/use-char-for-single-char-strings/src/colorize.cr):
there are still these single char strings:
```Crystal
  private MODE_DEFAULT   = "0"
  private MODE_BOLD      = "1"
  private MODE_BRIGHT    = "1"
  private MODE_DIM       = "2"
  private MODE_UNDERLINE = "4"
  private MODE_BLINK     = "5"
  private MODE_REVERSE   = "7"
  private MODE_HIDDEN    = "8"
```
shouldn't they be better changed to
```Crystal
  private MODE_DEFAULT   = '0'
  private MODE_BOLD      = '1'
  private MODE_BRIGHT    = '1'
  private MODE_DIM       = '2'
  private MODE_UNDERLINE = '4'
  private MODE_BLINK     = '5'
  private MODE_REVERSE   = '7'
  private MODE_HIDDEN    = '8'
```

**Comment by Sija** [comment]
@r00ster91 I wonder why they are strings instead of ints anyway?

**Comment by r00ster91** [comment]
@Sija Hmmm I guess because then you don't need to convert it to an String first when you use it in this macro
```Crystal
        {% for name in %w(bold dim underline blink reverse hidden) %}
          if @mode.bits_set? MODE_{{name.upcase.id}}_FLAG
            io << ';' if printed
            io << MODE_{{name.upcase.id}}
            printed = true
          end
        {% end %}
```

**Comment by Sija** [comment]
@r00ster91 right, I've force-pushed changes.

File: src/compiler/crystal/semantic/cleanup_transformer.cr [hunk 2]
```diff
@@ -461,11 +461,8 @@ module Crystal
 
     def untyped_expression(node, msg = nil)
       ex_msg = String.build do |str|
-        str << "can't execute `"
-        str << node
-        str << "`"
-        str << " at "
-        str << node.location
+        str << "can't execute `" << node << '`'
+        str << " at " << node.location
```
Comment (RX14) [inline src/compiler/crystal/semantic/cleanup_transformer.cr]:
```cr
str << "can't execute `" << node << "` at " << node.location
```

File: src/slice.cr [hunk 3]
```diff
@@ -479,15 +479,15 @@ struct Slice(T)
   def to_s(io)
     if T == UInt8
       io << "Bytes"
-      io << "["
+      io << '['
```
Comment (j8r) [inline src/slice.cr]: Merge both strings to `"Bytes["`

File: src/slice.cr [hunk 4]
```diff
@@ -479,15 +479,15 @@ struct Slice(T)
...
       # Inspect using to_s because we know this is a UInt8.
       join ", ", io, &.to_s(io)
-      io << "]"
+      io << ']'
     else
       io << "Slice"
-      io << "["
+      io << '['
```
Comment (j8r) [inline src/slice.cr]: Ditto, `"Slice["`

**Comment by Sija** [comment]
Woot, another exception...

```
_main.o: In function `__crystal_main':

main_module:(.text+0x7aa): undefined reference to `*String::Builder@IO#<<<Foo+:Class>:String::Builder'

/usr/bin/ld: /tmp/crystal/crystal-run-crystal-spec-output.tmp: No symbol version section for versioned symbol `*String::Builder@IO#<<<Foo+:Class>:String::Builder'

/usr/bin/ld: final link failed: Nonrepresentable section on output

collect2: error: ld returned 1 exit status

Error: execution of command failed with code: 1: `cc "${@}" -o '/tmp/crystal/crystal-run-crystal-spec-output.tmp'  -rdynamic  -lpcre -lgc -lpthread /mnt/src/ext/libcrystal.a -levent -lrt -ldl -L/usr/lib -L/usr/local/lib`
```

**Comment by RX14** [comment]
I think it's just some weird race-condition/bug...

File: src/slice.cr (hunk 3 above)
Comment (Sija) [reply src/slice.cr]: done & force-pushed.

**Pull request merged**